benches/parse.rs

```diff
@@ -22,101 +22,72 @@ Keep-Alive: 115\r\n\
+HTTP/1.1 200 OK\r\n\
+Date: Wed, 21 Oct 2015 07:28:00 GMT\r\n\
+Host: www.kittyhell.com\r\n\
+User-Agent: Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10.6; ja-JP-mac; rv:1.9.2.3) Gecko/20100401 Firefox/3.6.3 Pathtraq/0.9\r\n\
+Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\r\n\
+Accept-Language: ja,en-us;q=0.7,en;q=0.3\r\n\
+Accept-Encoding: gzip,deflate\r\n\
+Accept-Charset: Shift_JIS,utf-8;q=0.7,*;q=0.7\r\n\
+Keep-Alive: 115\r\n\
+Connection: keep-alive\r\n\
+Cookie: wp_ozh_wsa_visits=2; wp_ozh_wsa_visit_lasttime=xxxxxxxxxx; __utma=xxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.xxxxxxxxxx.x; __utmz=xxxxxxxxx.xxxxxxxxxx.x.x.utmccn=(referral)|utmcsr=reader.livedoor.com|utmcct=/reader/|utmcmd=referral|padding=under256\r\n\r\n";
 
-    #[repr(C)]
-    struct Headers<'a>(&'a mut [Header<'a>]);
-    let method = [0i8; 16];
-    let path = [0i8; 16];
-    let mut minor_version = 0;
-    let mut h = [Header(&[], &[]); 16];
-    let mut h_len = h.len();
-    let headers = Headers(&mut h);
-    let prev_buf_len = 0;
+fn resp(c: &mut Criterion) {
+    let mut headers = [httparse::Header{ name: "", value: &[] }; 16];
+    let mut resp = httparse::Response::new(&mut headers);
 
-    b.iter(|| {
-        let ret = unsafe {
-            pico::ffi::phr_parse_request(
-                REQ_SHORT.as_ptr() as *const _,
-                REQ_SHORT.len(),
-                &mut method.as_ptr(),
-                &mut 16,
-                &mut path.as_ptr(),
-                &mut 16,
-                &mut minor_version,
-                mem::transmute::<*mut Header, *mut pico::ffi::phr_header>(headers.0.as_mut_ptr()),
-                &mut h_len as *mut usize as *mut _,
-                prev_buf_len
-            )
-        };
-        assert_eq!(ret, REQ_SHORT.len() as i32);
-    });
-    b.bytes = REQ_SHORT.len() as u64;
+    c.benchmark_group("resp")
+        .throughput(Throughput::Bytes(RESP.len() as u64))
+        .bench_function("resp", |b| b.iter(|| {
+            assert_eq!(black_box(resp.parse(RESP).unwrap()), httparse::Status::Complete(RESP.len()));
+        }));
 }
 
-#[bench]
-fn bench_httparse_short(b: &mut test::Bencher) {
+fn resp_short(c: &mut Criterion) {
     let mut headers = [httparse::Header{ name: "", value: &[] }; 16];
-    let mut req = httparse::Request::new(&mut headers);
-    b.iter(|| {
-        assert_eq!(req.parse(REQ_SHORT).unwrap(), httparse::Status::Complete(REQ_SHORT.len()));
-    });
-    b.bytes = REQ_SHORT.len() as u64;
+    let mut resp = httparse::Response::new(&mut headers);
+
+    c.benchmark_group("resp_short")
+        .throughput(Throughput::Bytes(RESP_SHORT.len() as u64))
+        .bench_function("resp_short", |b| b.iter(|| {
+            assert_eq!(black_box(resp.parse(RESP_SHORT).unwrap()), httparse::Status::Complete(RESP_SHORT.len()));
+        }));
+}
+
+criterion_group!{
+    name = benches;
+    config = Criterion::default().sample_size(100).measurement_time(Duration::from_secs(10));
```

Comment:
This 10 second number can be tweaked if it slows down CI too much, but when spotting performance regressions for functions this small it's very helpful for keeping noise down.